src/LambdaBTree.h

```diff
@@ -0,0 +1,654 @@
+
+        // insert using iterative implementation
+
+        typename parenttype::node* cur = nullptr;
+
+        // test last insert hints
+        typename parenttype::lock_type::Lease cur_lease;
+
+        auto checkHint = [&](typename parenttype::node* last_insert) {
+            // ignore null pointer
+            if (!last_insert) return false;
+            // get a read lease on indicated node
+            auto hint_lease = last_insert->lock.start_read();
+            // check whether it covers the key
+            if (!this->weak_covers(last_insert, k)) return false;
+            // and if there was no concurrent modification
+            if (!last_insert->lock.validate(hint_lease)) return false;
+            // use hinted location
+            cur = last_insert;
+            // and keep lease
+            cur_lease = hint_lease;
+            // we found a hit
+            return true;
+        };
+
+        if (hints.last_insert.any(checkHint)) {
+            // register this as a hit
+            this->hint_stats.inserts.addHit();
+        } else {
+            // register this as a miss
+            this->hint_stats.inserts.addMiss();
+        }
+
+        // if there is no valid hint ..
+        if (!cur) {
+            do {
+                // get root - access lock
+                auto root_lease = this->root_lock.start_read();
+
+                // start with root
+                cur = this->root;
+
+                // get lease of the next node to be accessed
+                cur_lease = cur->lock.start_read();
+
+                // check validity of root pointer
+                if (this->root_lock.end_read(root_lease)) break;
+
+            } while (true);
+        }
+
+        while (true) {
+            // handle inner nodes
+            if (cur->inner) {
+                auto a = &(cur->keys[0]);
+                auto b = &(cur->keys[cur->numElements]);
+
+                auto pos = this->search.lower_bound(k, a, b, this->weak_comp);
+                auto idx = pos - a;
+
+                // early exit for sets
+                if (isSet && pos != b && this->weak_equal(*pos, k)) {
+                    // validate results
+                    if (!cur->lock.validate(cur_lease)) {
+                        // start over again
+                        return insert(k, hints, f);
+                    }
+
+                    // update provenance information
+                    if (typeid(Comparator) != typeid(WeakComparator) && this->less(k, *pos)) {
+                        if (!cur->lock.try_upgrade_to_write(cur_lease)) {
+                            // start again
+                            return insert(k, hints, f);
+                        }
+                        this->update(*pos, k);
+
+                        // get result before releasing lock
+                        auto res = (*pos).second;
+
+                        cur->lock.end_write();
+                        // XXX (pnappa): is this correct..?
```

Comment:
Is this XXX done? If these comments are staying then it looks like they should be TODOs